storage/inmem/inmem.go

```diff
@@ -0,0 +1,248 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package inmem
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/armon/go-radix"
+	"github.com/hashicorp/nodeenrollment"
+	"github.com/hashicorp/nodeenrollment/types"
+	"google.golang.org/protobuf/proto"
+)
+
+// These paths are used for different kinds of storage
+const (
+	rootsSubPath                 = "roots"
+	nodeInfoSubPath              = "nodeinfo"
+	nodeCredsSubPath             = "nodecreds"
+	serverLedActivationTokenPath = "serverledactivationtokens"
+)
+
+var _ nodeenrollment.Storage = (*Storage)(nil)
+
+// InmemBackend is an in-memory only physical backend. It is useful
+// for testing and development situations where the data is not
+// expected to be durable.
+type Storage struct {
+	sync.RWMutex
+	root *radix.Tree
+}
+
+// New creates a new object that implements the Storage interface in memory. It
+// is thread-safe.
+func New(ctx context.Context) (*Storage, error) {
+	return &Storage{
+		root: radix.New(),
+	}, nil
+}
+
+// subPathFromMsg determines what subdirectory to use based on which message
+// type it is
+func subPathFromMsg(msg proto.Message) (string, error) {
+	const op = "nodeenrollment.storage.inmem.(Storage).subPathFromMsg"
+	switch t := msg.(type) {
+	case *types.NodeCredentials:
+		return nodeCredsSubPath, nil
+	case *types.NodeInformation:
+		return nodeInfoSubPath, nil
+	case *types.RootCertificates:
+		return rootsSubPath, nil
+	case *types.ServerLedActivationToken:
+		return serverLedActivationTokenPath, nil
+	default:
+		return "", fmt.Errorf("(%s) unknown message type %T", op, t)
+	}
+}
+
+// Store implements the Storage interface.
+//
+// If the message already exists, it is overwritten.
+func (ts *Storage) Store(ctx context.Context, msg nodeenrollment.MessageWithId) error {
+	const op = "nodeenrollment.storage.inmem.(Storage).Store"
+	if err := types.ValidateMessage(msg); err != nil {
+		return fmt.Errorf("(%s) given message cannot be stored: %w", op, err)
+	}
+	subPath, err := subPathFromMsg(msg)
+	if err != nil {
+		return fmt.Errorf("(%s) given message cannot be stored: %w", op, err)
+	}
+	return ts.storeValue(ctx, msg.GetId(), subPath, msg)
+}
+
+// Load implements the Storage interface
+func (ts *Storage) Load(ctx context.Context, msg nodeenrollment.MessageWithId) error {
+	const op = "nodeenrollment.storage.inmem.(Storage).Load"
+	if err := types.ValidateMessage(msg); err != nil {
+		return fmt.Errorf("(%s) given message cannot be loaded: %w", op, err)
+	}
+	subPath, err := subPathFromMsg(msg)
+	if err != nil {
+		return fmt.Errorf("(%s) given message cannot be loaded: %w", op, err)
+	}
+	err = ts.loadValue(ctx, msg.GetId(), subPath, msg)
```

Comment:
perhaps add a test, where the msg is not found?

Reply:
Added